wallet/chainntfns.go

```diff
@@ -329,3 +344,151 @@ func (w *Wallet) addRelevantTx(dbtx walletdb.ReadWriteTx, rec *wtxmgr.TxRecord,
 
 	return nil
 }
+
+// birthdaySanityCheck is a helper function that ensures our birthday block
+// correctly reflects the birthday timestamp within a reasonable timestamp
+// delta. It will be run after the wallet establishes its connection with the
+// backend, but before it begins syncing.  This is done as the second part to
+// the wallet's address manager migration where we populate the birthday block
+// to ensure we do not miss any relevant events throughout rescans.
+func (w *Wallet) birthdaySanityCheck() error {
+	// We'll start by acquiring our chain backend client as we'll be
+	// querying it for blocks.
+	chainClient, err := w.requireChainClient()
+	if err != nil {
+		return err
+	}
+
+	// We'll then fetch our wallet's birthday timestamp and block.
+	birthdayTimestamp := w.Manager.Birthday()
+	var birthdayBlock waddrmgr.BlockStamp
+	err = walletdb.View(w.db, func(tx walletdb.ReadTx) error {
+		var err error
+		ns := tx.ReadBucket(waddrmgrNamespaceKey)
+		birthdayBlock, err = w.Manager.BirthdayBlock(ns)
+		return err
+	})
+
+	switch {
+	// If our wallet's birthday block has not been set yet, then this is our
+	// initial sync, so we'll defer setting it until then.
+	case waddrmgr.IsError(err, waddrmgr.ErrBirthdayBlockNotSet):
+		return nil
+
+	// Otherwise, we'll return the error if there was one.
+	case err != nil:
+		return err
+	}
+
+	log.Debugf("Starting sanity check for the wallet's birthday block "+
+		"from: height=%d, hash=%v", birthdayBlock.Height,
+		birthdayBlock.Hash)
+
+	// Now, we'll need to determine if our block correctly reflects our
+	// timestamp. To do so, we'll fetch the block header and check its
+	// timestamp.
+	candidate := birthdayBlock
+	header, err := chainClient.GetBlockHeader(&candidate.Hash)
+	if err != nil {
+		return fmt.Errorf("unable to get header for block hash %v: %v",
+			candidate.Hash, err)
+	}
+
+	// We'll go back a day worth of blocks in the chain until we find a
+	// block whose timestamp is below our birthday timestamp.
+	heightDelta := int32(144)
+	for birthdayTimestamp.Before(header.Timestamp) {
```

Comment:
Since header timestamps are not necessarily increasing, we must be carful here. I'm not sure if there's a faster way of finding the first block with a timestamp after our birthday without scanning all blocks, but I think we can make a decent estimate.

Since a block's timestamp must be greater than the [median of the previous 11 blocks](https://en.bitcoin.it/wiki/Block_timestamp), a reasonable strategy would be to 
* find the last block with a timestamp before our birthday
* take the maximum of the previous 11 block window
* keep moving this window back until the maximum is < birthday

This lets us find the last 11 block window with all blocks before the birthday, and the next block (together with a delta, see below) should be an okay guess I think.

Reply:
Not sure if all this additional complexity is warranted for since the birthday timestamp is set to two days in the past from when it actually is. Also, `birthdaySanityCheck` only serves useful for existing wallets that have gone through the birthday block migration as the assumption used within it (1 block/10 mins) is most likely to be inaccurate. For new wallets, the correct block will be populated upon syncing with the chain.